spanhelper: use global logrus hook and a custom formatter

xzfc opened the pull request:
To avoid data-races with `logrus.SetLevel()` used in tests.
`GetLevel()` uses atomic read internally.

denis-tingajkin commented:
> To avoid data-races with logrus.SetLevel() 

@xzfc Could you explain how we can face the race condition?

xzfc commented:
Just got this warning once while calling `SetLevel()` at start of the test.

https://github.com/xzfc/networkservicemesh-sdk/blob/c9f2d35db82d9827c57763f32d3392126aea4aa8/pkg/networkservice/common/refresh/client_test.go#L206-L208

<details><summary>log</summary>

```
2020-11-25T11:32:54.4226040Z === RUN   TestRefreshClient_Sandbox
2020-11-25T11:32:54.4226950Z ==================
2020-11-25T11:32:54.4227630Z WARNING: DATA RACE
2020-11-25T11:32:54.4231360Z Write at 0x00c00022e10c by goroutine 209:
2020-11-25T11:32:54.4234740Z   sync/atomic.StoreInt32()
2020-11-25T11:32:54.4252250Z       /Users/runner/hostedtoolcache/go/1.15.5/x64/src/runtime/race_amd64.s:229 +0xb
2020-11-25T11:32:54.4254480Z   github.com/sirupsen/logrus.(*Logger).SetLevel()
2020-11-25T11:32:54.4255710Z       /Users/runner/go/pkg/mod/github.com/sirupsen/logrus@v1.7.0/logger.go:357 +0x11b
2020-11-25T11:32:54.4256750Z   github.com/sirupsen/logrus.SetLevel()
2020-11-25T11:32:54.4257840Z       /Users/runner/go/pkg/mod/github.com/sirupsen/logrus@v1.7.0/exported.go:36 +0xee
2020-11-25T11:32:54.4259270Z   github.com/networkservicemesh/sdk/pkg/networkservice/common/refresh_test.TestRefreshClient_Sandbox()
2020-11-25T11:32:54.4262330Z       /Users/runner/work/networkservicemesh-sdk/networkservicemesh-sdk/pkg/networkservice/common/refresh/client_test.go:208 +0xed
2020-11-25T11:32:54.4263660Z   testing.tRunner()
2020-11-25T11:32:54.4264660Z       /Users/runner/hostedtoolcache/go/1.15.5/x64/src/testing/testing.go:1123 +0x202
2020-11-25T11:32:54.4265470Z 
2020-11-25T11:32:54.4287820Z Previous read at 0x00c00022e10c by goroutine 22:
2020-11-25T11:32:54.4289340Z   github.com/networkservicemesh/sdk/pkg/tools/spanhelper.LogFromSpan()
2020-11-25T11:32:54.4291760Z       /Users/runner/work/networkservicemesh-sdk/networkservicemesh-sdk/pkg/tools/spanhelper/span_helper.go:109 +0x1b5
2020-11-25T11:32:54.4293410Z   github.com/networkservicemesh/sdk/pkg/tools/spanhelper.(*spanHelper).Logger()
2020-11-25T11:32:54.4295640Z       /Users/runner/work/networkservicemesh-sdk/networkservicemesh-sdk/pkg/tools/spanhelper/span_helper.go:235 +0xcd
2020-11-25T11:32:54.4309410Z   github.com/networkservicemesh/sdk/pkg/registry/core/trace.(*traceNetworkServiceEndpointRegistryServer).Find()
2020-11-25T11:32:54.4314570Z       /Users/runner/work/networkservicemesh-sdk/networkservicemesh-sdk/pkg/registry/core/trace/nse_registry.go:171 +0x194
2020-11-25T11:32:54.4323990Z   github.com/networkservicemesh/sdk/pkg/registry/core/next.(*nextNetworkServiceEndpointRegistryServer).Find()
2020-11-25T11:32:54.4334520Z       /Users/runner/work/networkservicemesh-sdk/networkservicemesh-sdk/pkg/registry/core/next/nse_registry_server.go:62 +0x3cd
2020-11-25T11:32:54.4338000Z   github.com/networkservicemesh/sdk/pkg/registry/memory.(*networkServiceEndpointRegistryServer).Find()
2020-11-25T11:32:54.4340540Z       /Users/runner/work/networkservicemesh-sdk/networkservicemesh-sdk/pkg/registry/memory/nse_server.go:107 +0x581
2020-11-25T11:32:54.4349450Z   github.com/networkservicemesh/sdk/pkg/registry/core/trace.(*traceNetworkServiceEndpointRegistryServer).Find()
2020-11-25T11:32:54.4352460Z       /Users/runner/work/networkservicemesh-sdk/networkservicemesh-sdk/pkg/registry/core/trace/nse_registry.go:178 +0x326
2020-11-25T11:32:54.4357950Z   github.com/networkservicemesh/sdk/pkg/registry/core/next.(*nextNetworkServiceEndpointRegistryServer).Find()
2020-11-25T11:32:54.4360470Z       /Users/runner/work/networkservicemesh-sdk/networkservicemesh-sdk/pkg/registry/core/next/nse_registry_server.go:62 +0x3cd
2020-11-25T11:32:54.4364180Z   github.com/networkservicemesh/sdk/pkg/registry/common/expire.(*nseServer).Find()
2020-11-25T11:32:54.4366500Z       /Users/runner/work/networkservicemesh-sdk/networkservicemesh-sdk/pkg/registry/common/expire/nse_server.go:57 +0x8c
2020-11-25T11:32:54.4368510Z   github.com/networkservicemesh/sdk/pkg/registry/core/trace.(*traceNetworkServiceEndpointRegistryServer).Find()
2020-11-25T11:32:54.4370930Z       /Users/runner/work/networkservicemesh-sdk/networkservicemesh-sdk/pkg/registry/core/trace/nse_registry.go:178 +0x326
2020-11-25T11:32:54.4384650Z   github.com/networkservicemesh/sdk/pkg/registry/core/next.(*nextNetworkServiceEndpointRegistryServer).Find()
2020-11-25T11:32:54.4387960Z       /Users/runner/work/networkservicemesh-sdk/networkservicemesh-sdk/pkg/registry/core/next/nse_registry_server.go:62 +0x3cd
2020-11-25T11:32:54.4434770Z   github.com/networkservicemesh/sdk/pkg/registry/common/setid.(*setIDNetworkServiceEndpointRegistryServer).Find()
2020-11-25T11:32:54.4466290Z       /Users/runner/work/networkservicemesh-sdk/networkservicemesh-sdk/pkg/registry/common/setid/network_service_registry_server.go:48 +0xe1
2020-11-25T11:32:54.4499670Z   github.com/networkservicemesh/sdk/pkg/registry/core/trace.(*traceNetworkServiceEndpointRegistryServer).Find()
2020-11-25T11:32:54.4550220Z       /Users/runner/work/networkservicemesh-sdk/networkservicemesh-sdk/pkg/registry/core/trace/nse_registry.go:178 +0x326
2020-11-25T11:32:54.4594180Z   github.com/networkservicemesh/sdk/pkg/registry/core/next.(*nextNetworkServiceEndpointRegistryServer).Find()
2020-11-25T11:32:54.4644360Z       /Users/runner/work/networkservicemesh-sdk/networkservicemesh-sdk/pkg/registry/core/next/nse_registry_server.go:62 +0x3cd
2020-11-25T11:32:54.4693720Z   github.com/networkservicemesh/sdk/pkg/registry/core/next.(*nextNetworkServiceEndpointRegistryServer).Find()
2020-11-25T11:32:54.4781460Z       /Users/runner/work/networkservicemesh-sdk/networkservicemesh-sdk/pkg/registry/core/next/nse_registry_server.go:62 +0x3cd
2020-11-25T11:32:54.4837990Z   github.com/networkservicemesh/sdk/pkg/registry/core/adapters.nseFindServerToClient.func1()
2020-11-25T11:32:54.4848410Z       /Users/runner/work/networkservicemesh-sdk/networkservicemesh-sdk/pkg/registry/core/adapters/nse_registry.go:128 +0xb8
2020-11-25T11:32:54.4854050Z 
2020-11-25T11:32:54.4859410Z Goroutine 209 (running) created at:
2020-11-25T11:32:54.4863050Z   testing.(*T).Run()
2020-11-25T11:32:54.4868470Z       /Users/runner/hostedtoolcache/go/1.15.5/x64/src/testing/testing.go:1168 +0x5bb
2020-11-25T11:32:54.4871220Z   testing.runTests.func1()
2020-11-25T11:32:54.4872430Z       /Users/runner/hostedtoolcache/go/1.15.5/x64/src/testing/testing.go:1439 +0xa6
2020-11-25T11:32:54.4873400Z   testing.tRunner()
2020-11-25T11:32:54.4874530Z       /Users/runner/hostedtoolcache/go/1.15.5/x64/src/testing/testing.go:1123 +0x202
2020-11-25T11:32:54.4875490Z   testing.runTests()
2020-11-25T11:32:54.4876540Z       /Users/runner/hostedtoolcache/go/1.15.5/x64/src/testing/testing.go:1437 +0x612
2020-11-25T11:32:54.4880950Z   testing.(*M).Run()
2020-11-25T11:32:54.4882010Z       /Users/runner/hostedtoolcache/go/1.15.5/x64/src/testing/testing.go:1345 +0x3b3
2020-11-25T11:32:54.4882900Z   main.main()
2020-11-25T11:32:54.4883620Z       _testmain.go:53 +0x236
2020-11-25T11:32:54.4884170Z 
2020-11-25T11:32:54.4885100Z Goroutine 22 (finished) created at:
2020-11-25T11:32:54.4886820Z   github.com/networkservicemesh/sdk/pkg/registry/core/adapters.nseFindServerToClient()
2020-11-25T11:32:54.4889460Z       /Users/runner/work/networkservicemesh-sdk/networkservicemesh-sdk/pkg/registry/core/adapters/nse_registry.go:126 +0x1b4
2020-11-25T11:32:54.4891810Z   github.com/networkservicemesh/sdk/pkg/registry/core/adapters.(*networkServiceEndpointRegistryClient).Find()
2020-11-25T11:32:54.4893990Z       /Users/runner/work/networkservicemesh-sdk/networkservicemesh-sdk/pkg/registry/core/adapters/nse_registry.go:85 +0x178
2020-11-25T11:32:54.4895670Z   github.com/networkservicemesh/sdk/pkg/registry/common/expire.(*nsServer).checkUpdates()
2020-11-25T11:32:54.4897760Z       /Users/runner/work/networkservicemesh-sdk/networkservicemesh-sdk/pkg/registry/common/expire/ns_server.go:43 +0x203
2020-11-25T11:32:54.4899460Z   github.com/networkservicemesh/sdk/pkg/registry/common/expire.(*nsServer).Register.func1.1()
2020-11-25T11:32:54.4901570Z       /Users/runner/work/networkservicemesh-sdk/networkservicemesh-sdk/pkg/registry/common/expire/ns_server.go:81 +0x3e
2020-11-25T11:32:54.4902790Z ==================
```

</details>

Full log: https://github.com/xzfc/networkservicemesh-sdk/runs/1453161814?check_suite_focus=true

This one probably is a false-positive detection (since the reading goroutine is finished), but I can think of a potential legitimate use-case. And it's a good style to use a provided getter anyway.

edwarnicke commented:
Does it make sense to use Getters for the other values as well?

xzfc commented:
> Does it make sense to use Getters for the other values as well?

There is only one getter method in Logger ([godoc](https://godoc.org/github.com/sirupsen/logrus#Logger.GetLevel)). Probably, it's the only field intended to be configured at arbitrary time point.

denis-tingajkin commented:
I feel you can also try to set any public field of struct https://github.com/sirupsen/logrus/blob/master/logger.go#L17 in start of your test(https://github.com/xzfc/networkservicemesh-sdk/blob/c9f2d35db82d9827c57763f32d3392126aea4aa8/pkg/networkservice/common/refresh/client_test.go#L208) and face the same issue. So  this patch is not fix a root cause. 

So It seems to me the root cause of the problem can be related to your testing apporoch. Could you attach the command how you run the test?

xzfc commented:
> Could you attach the command how you run the test?

`${GOPATH}/bin/gotestsum -f standard-verbose --junitfile ~/junit/unit-tests.xml -- -race github.com/networkservicemesh/sdk/pkg/networkservice/common/refresh -count=10 -run '^TestRefreshClient_Sandbox$'`

https://github.com/xzfc/networkservicemesh-sdk/blob/c9f2d35db82d9827c57763f32d3392126aea4aa8/.github/workflows/issue-260.yaml#L51

denis-tingajkin commented:
I guess the problem is related to parameter `count=10`, also I mostly sure that you have goroutine leaks. Lets run the tests without parameter `goleak.IgnoreCurrent()` and lets fix find and fix leaks.

xzfc commented:
The root of the problem: the default logrus logger is a shared state. Function `LogFromSpan` reads the shared state without any synchronization, and various tests update this state (via `logrus.SetLevel()` or `logrus.SetOutput()`). Hence the potential data race. This situation could be simplified to this test code, which also triggers the data race detector.

<details><summary>test code</summary>

```go
package main_test

import (
	"context"
	"fmt"
	"sync/atomic"
	"testing"
	"time"
)

var LogLevel int32

func TestFoo(t *testing.T) {
	ctx, cancel := context.WithCancel(context.Background())

	// Similar to
	// https://github.com/networkservicemesh/sdk/blob/f3c3c51c88e2456011d1a6075a8556d768574c4c/pkg/registry/common/expire/ns_server.go#L79-L83
	go func() {
		for ctx.Err() == nil {
			lvl := LogLevel // Similar to `logrus.StandardLogger().Level`
			if lvl >= 1 {
				fmt.Println(".")
			}
			time.Sleep(100 * time.Millisecond)
		}
	}()

	time.Sleep(time.Second)
	cancel()
	time.Sleep(time.Second)
}

func TestBar(t *testing.T) {
	atomic.StoreInt32(&LogLevel, 1) // Similar to `logrus.SetLevel()`
}
```

</details>



<details><summary>go test -v -race</summary>

```
=== RUN   TestFoo
--- PASS: TestFoo (2.00s)
=== RUN   TestBar
==================
WARNING: DATA RACE
Write at 0x000000764064 by goroutine 14:
  sync/atomic.StoreInt32()
      /nix/store/bqdbizzgdcckkci0splpnhgy1c13h2z0-go-1.15.3/share/go/src/runtime/race_amd64.s:229 +0xb
  example.com/m_test.TestBar()
      /x/Experiments/go/2020-11-25-leak/main_test.go:34 +0x44
  testing.tRunner()
      /nix/store/bqdbizzgdcckkci0splpnhgy1c13h2z0-go-1.15.3/share/go/src/testing/testing.go:1123 +0x202

Previous read at 0x000000764064 by goroutine 8:
  example.com/m_test.TestFoo.func1()
      /x/Experiments/go/2020-11-25-leak/main_test.go:20 +0x6d

Goroutine 14 (running) created at:
  testing.(*T).Run()
      /nix/store/bqdbizzgdcckkci0splpnhgy1c13h2z0-go-1.15.3/share/go/src/testing/testing.go:1168 +0x5bb
  testing.runTests.func1()
      /nix/store/bqdbizzgdcckkci0splpnhgy1c13h2z0-go-1.15.3/share/go/src/testing/testing.go:1439 +0xa6
  testing.tRunner()
      /nix/store/bqdbizzgdcckkci0splpnhgy1c13h2z0-go-1.15.3/share/go/src/testing/testing.go:1123 +0x202
  testing.runTests()
      /nix/store/bqdbizzgdcckkci0splpnhgy1c13h2z0-go-1.15.3/share/go/src/testing/testing.go:1437 +0x612
  testing.(*M).Run()
      /nix/store/bqdbizzgdcckkci0splpnhgy1c13h2z0-go-1.15.3/share/go/src/testing/testing.go:1345 +0x3b3
  main.main()
      _testmain.go:47 +0x236

Goroutine 8 (finished) created at:
  example.com/m_test.TestFoo()
      /x/Experiments/go/2020-11-25-leak/main_test.go:18 +0x7c
  testing.tRunner()
      /nix/store/bqdbizzgdcckkci0splpnhgy1c13h2z0-go-1.15.3/share/go/src/testing/testing.go:1123 +0x202
==================
    testing.go:1038: race detected during execution of test
--- FAIL: TestBar (0.01s)
=== CONT  
    testing.go:1038: race detected during execution of test
FAIL
exit status 1
FAIL	example.com/m	2.015s
```

</details>

There is no real goroutine leakage, and no race condition occurred. But the race detector is too picky and reports a potential data race (which may happen for real if we remove the last 1 second sleep).

---

I've updated the PR. New description:
Before this PR, the function [`LogFromSpan`] cloned `StandardLogger` only to add a new trace hook with a `span`.
In this commit, the trace hook is enabled globally and takes `span` from a provided context.
Now, there are no shared logrus state reads in our code.

Also, in this commit, the global logrus formatter isn't used for jaeger log entries to achieve a more laconic output. Removed information that is already present in jaeger (time, operation, span) and terminal color escape codes. Examples of the same entry:
```
# old formatter (nested-logrus-formatter as used in cmd-nsmgr):
Nov 27 00:29:14.019[36m [INFO] [operation:sdk/pkg/networkservice/common/timeout/timeoutServer.Request] [span:353b7f5455bac1d7:755e29a4ec7bd0bf:0f192d62140cb404:1] [0msomething

# new formatter:
[INFO] something
```

[`LogFromSpan`]: https://github.com/networkservicemesh/sdk/blob/f3c3c51c88e2456011d1a6075a8556d768574c4c/pkg/tools/spanhelper/span_helper.go#L104-L122

denis-tingajkin commented on pkg/tools/spanhelper/span_helper.go:
```diff
@@ -102,48 +108,63 @@ func fromContext(ctx context.Context) *traceCtxInfo {
+	return logrus.StandardLogger().
+		WithField(spanField, span).
+		WithContext(context.WithValue(
+			context.Background(),
```
Should we use here span.Context() in case if the span is not nil?

denis-tingaikin commented:
This is not actually more.

The problem solved in 
https://github.com/networkservicemesh/sdk/pull/680
https://github.com/networkservicemesh/sdk/pull/641
